### PR TITLE
Allow user to specify git clone directory in gitRepo volume

### LINKED ISSUE #15421: Support renaming clone in gitRepo volumes
From [stackoverflow thread](http://stackoverflow.com/questions/33044198/naming-gitrepo-mount-path-in-kubernetes), a user would like to set the destination name of the cloned repository in a gitRepo volume.

`git clone` supports supplying a destination name for a cloned repository, e.g. `git clone https://github.com/kubernetes/kubernetes.git k8s`, so I think it is reasonable for us to support the same.

### PR DESCRIPTION
**Pull request by resouer**
Fix #15421

### CODE REVIEW

**Comment by ncdc** [comment]
@kubernetes/rh-cluster-infra @kubernetes/rh-storage @kubernetes/rh-dev-experience @kubernetes/rh-ux

File: pkg/volume/git_repo/git_repo.go [hunk 1]
```diff
@@ -138,8 +140,15 @@ func (b *gitRepoVolumeBuilder) SetUpAt(dir string) error {
...
+	// If provided target directory
+	if "" != b.target {
+		if output, err := b.execCommand("git", []string{"clone", b.source, b.target}, dir); err != nil {
+			return fmt.Errorf("failed to exec 'git clone %s %s': %s: %v", b.source, b.target, output, err)
+		}
+	} else {
+		if output, err := b.execCommand("git", []string{"clone", b.source}, dir); err != nil {
+			return fmt.Errorf("failed to exec 'git clone %s': %s: %v", b.source, output, err)
+		}
 	}
```
Comment (markturansky) [inline pkg/volume/git_repo/git_repo.go]: More concise this way and also supports N git args in the error message:

```
	args := []string{"clone", b.source}
	if b.target != "" {
		args = append(args, b.target)
	}
	if output, err := b.execCommand("git", args, dir); err != nil {
		return fmt.Errorf("failed to exec 'git %s': %s: %v", strings.Join(append(args[:1], args[1:len(args)]...), " "), output, err)
	}
```

Comment (markturansky) [inline pkg/volume/git_repo/git_repo.go]: Will this part fail if you specify a target?  The current implementation assumes there will be exactly 1 directory (the repo name) in the volume.

```
if len(files) != 1 {
 		return fmt.Errorf("unexpected directory contents: %v", files)
 	}
```

I couldn't comment on that line in the diff because you didn't edit it.

Comment (resouer) [inline pkg/volume/git_repo/git_repo.go]: This will fail when clone directory is '.'

Actually, see my comments here: https://github.com/kubernetes/kubernetes/issues/15421#issuecomment-147963960

I'd like to update the code if we reached a agreement.

Comment (thockin) [inline pkg/volume/git_repo/git_repo.go]: As per #15421

How urgent is this one small feature? I'm not deeply against adding it if
it solves real problems, but I'd rather focus energy on the git-sync
sidecar. Additionally, we need clarity on whether this is just "checkout
and rename" or whether "." is expected to be supported. Doing a checkout to
a subdir has one significant advantage - it can be updated atomically. This
means you could periodically sync to head of a branch, which seems like a
powerful feature to me.

If we support ".", that does not work.

**Comment by resouer** [comment]
@markturansky Although discusss is in progess in original issue, I updated this with tests

**Comment by resouer** [comment]
`serialization_test.go` continually fails unless I run `./hack/update-codecgen.sh`, which leads to a huge change ... don't know how to solve this for now ...

File: pkg/volume/git_repo/git_repo.go [hunk 2]
```diff
@@ -138,24 +141,40 @@ func (b *gitRepoVolumeBuilder) SetUpAt(dir string) error {
...
+	if b.target == "." {
+		// if target dir is '.', use the current dir
+		subdir = path.Join(dir)
+	} else {
+		// else use the subdir
+		if len(files) != 1 {
+			return fmt.Errorf("unexpected directory contents: %v", files)
+		}
+		subdir = path.Join(dir, files[0].Name())
+	}
```
Comment (markturansky) [inline pkg/volume/git_repo/git_repo.go]: Better to use switch over if/else:

```
switch {
case b.target == "":
case b.target != "" && len(files) == 1:
case b.target != && len(files) != 1
}
```

File: pkg/volume/git_repo/git_repo.go [hunk 3]
```diff
@@ -138,24 +141,39 @@ func (b *gitRepoVolumeBuilder) SetUpAt(dir string) error {
...
 
 	files, err := ioutil.ReadDir(dir)
 	if err != nil {
 		return err
 	}
-	if len(files) != 1 {
-		return fmt.Errorf("unexpected directory contents: %v", files)
-	}
+
 	if len(b.revision) == 0 {
```
Comment (markturansky) [inline pkg/volume/git_repo/git_repo.go]: ` len(b.revision) == 0` and `b.target != ""` are the same (in effect) but different implementations.  They should be consistent.  I've mostly seen (and used) empty string "" comparison instead of string length check.

File: pkg/volume/git_repo/git_repo_test.go [hunk 4]
```diff
@@ -95,13 +103,26 @@ func testSetUp(plug volume.VolumePlugin, builder volume.Builder, t *testing.T) {
...
 		t.Errorf("unexpected commands: %v, expected: %v", fcmd.CombinedOutputLog, expectedCmds)
 	}
-	expectedDirs := []string{g.GetPath(), g.GetPath() + "/kubernetes", g.GetPath() + "/kubernetes"}
+
+	var repoDir string
+	if g.target == "." {
+		repoDir = ""
+	} else {
+		repoDir = "/kubernetes"
+	}
```
Comment (markturansky) [inline pkg/volume/git_repo/git_repo_test.go]: No need for `else`.

```
repoDir = ""
if g.target != "."{
   repoDir = "whatever"
}
```

**Comment by markturansky** [comment]
What kind of validation is required for the target directory?  "." and a path are valid.  Is "../path" valid? numbers/spaces/special characters?

What's the worst someone can put into the target string?

**Comment by resouer** [comment]
@markturansky I don't think numbers/special characters are harmful, for now.
I can see path abusing will be a hidden trouble, for example, clone to any places out of the volume path.
So, need to restrict `directory` to:
1. relative path
2. not '..'
3. no spaces

Or, we can only allow: [a-z][0-9][.], and refuse others

**Comment by lvlv** [comment]
Given these restrictions you suggested, does it make more sense as a
general volume feature instead of gitRepo only? e.g. we can add an
"InVolumePath" option in VolumeMount, so a container can mount a
sub-directory of a volume.

**Comment by liggitt** [comment]
For validation, should probably use path.Clean and compare to the original rather than trying to roll our own backstep or non-normalized detection

**Comment by pmorie** [comment]
@resouer @lvlv @liggitt @markturansky similar aspects of the validation problem for paths are dealt with in the downward API volume plugin, suggest taking a look there.

**Comment by resouer** [comment]
@pmorie @liggitt  Thanks for your advices

**Comment by thockin** [comment]
I'm still not convinced this change is worth the energy, rather than fixing
up the git-sync sidecar into something awesome.

**Comment by resouer** [comment]
@thockin I'd like to do, honestly. But I'm really not sure how. I fired up a issue #16247 and involve you in, we can list what we can do, and I will close this PR and start from there. 

What I try to avoid is using a half-baked demo to replace a build-in volume. Users will complain, I'm afraid ...

File: pkg/api/validation/validation.go [hunk 5]
```diff
@@ -440,9 +440,26 @@ func validateHostPathVolumeSource(hostPath *api.HostPathVolumeSource) errs.Valid
...
 	}
+	if path.IsAbs(target) {
+		allErrs = append(allErrs, errs.NewFieldInvalid("directory", target, "must not be an absolute path"))
+	}
+	items := strings.Split(target, string(os.PathSeparator))
+	for _, item := range items {
+		if item == ".." {
+			allErrs = append(allErrs, errs.NewFieldInvalid("directory", target, "must not contain \"..\""))
+		}
+		if item == " " {
```
Comment (markturansky) [inline pkg/api/validation/validation.go]: what about trim and a check for empty string?  that would find N spaces, not just 1.

File: pkg/api/validation/validation.go [hunk 6]
```diff
@@ -440,9 +440,26 @@ func validateHostPathVolumeSource(hostPath *api.HostPathVolumeSource) errs.Valid
...
+			allErrs = append(allErrs, errs.NewFieldInvalid("directory", target, "must not contain spaces"))
+		}
+	}
+	if strings.HasPrefix(items[0], "..") && len(items[0]) > 2 {
+		allErrs = append(allErrs, errs.NewFieldInvalid("directory", target, "must not start with \"..\""))
+	}
```
Comment (markturansky) [inline pkg/api/validation/validation.go]: Can this be rolled into the previous check for ".."?  One check for the same thing should be sufficient.

File: pkg/api/validation/validation_test.go [hunk 7]
```diff
@@ -560,6 +565,11 @@ func TestValidateVolumes(t *testing.T) {
 		"empty wwn":                  {[]api.Volume{{Name: "badimage", VolumeSource: zeroWWN}}, errors.ValidationErrorTypeRequired, "[0].source.fc.targetWWNs", ""},
 		"empty lun":                  {[]api.Volume{{Name: "badimage", VolumeSource: emptyLun}}, errors.ValidationErrorTypeRequired, "[0].source.fc.lun", ""},
 		"slash in datasetName":       {[]api.Volume{{Name: "slashinname", VolumeSource: slashInName}}, errors.ValidationErrorTypeInvalid, "[0].source.flocker.datasetName", "must not contain '/'"},
+		"starts with '..'":           {[]api.Volume{{Name: "badprefix", VolumeSource: startsWithDots}}, errors.ValidationErrorTypeInvalid, "[0].source.gitRepo.directory", "must not start with \"..\""},
+		"contains '..'":              {[]api.Volume{{Name: "containsdots", VolumeSource: containsDots}}, errors.ValidationErrorTypeInvalid, "[0].source.gitRepo.directory", "must not contain \"..\""},
+		"contains space:":            {[]api.Volume{{Name: "containsspace", VolumeSource: containsSpace}}, errors.ValidationErrorTypeInvalid, "[0].source.gitRepo.directory", "must not contain spaces"},
+
```
Comment (markturansky) [inline pkg/api/validation/validation_test.go]: dead line

File: pkg/volume/git_repo/git_repo_test.go [hunk 8]
```diff
@@ -78,14 +78,22 @@ func testSetUp(plug volume.VolumePlugin, builder volume.Builder, t *testing.T) {
 		},
 	}
 	g := builder.(*gitRepoVolumeBuilder)
+
```
Comment (markturansky) [inline pkg/volume/git_repo/git_repo_test.go]: dead line

**Comment by resouer** [comment]
@markturansky updated, thank you so much for your time
@ae6rt you can track the status of this PR

**Comment by pmorie** [comment]
@resouer Rebase needed

**Comment by resouer** [comment]
@pmorie Thanks. Rebase done

**Comment by markturansky** [comment]
LGTM.

**Comment by resouer** [comment]
@brendandburns any thoughts on gitrepo target?

File: pkg/api/validation/validation.go [hunk 9]
```diff
@@ -441,9 +441,27 @@ func validateHostPathVolumeSource(hostPath *api.HostPathVolumeSource) errs.Valid
 
 func validateGitRepoVolumeSource(gitRepo *api.GitRepoVolumeSource) errs.ValidationErrorList {
 	allErrs := errs.ValidationErrorList{}
+	target := gitRepo.Directory
+
 	if gitRepo.Repository == "" {
 		allErrs = append(allErrs, errs.NewFieldRequired("repository"))
 	}
+	if path.IsAbs(target) {
```
Comment (brendandburns) [inline pkg/api/validation/validation.go]:
```go
if len(target) == 0 {
  return allErrs
}
```

File: pkg/volume/git_repo/git_repo.go [hunk 10]
```diff
@@ -146,24 +149,39 @@ func (b *gitRepoVolumeBuilder) SetUpAt(dir string) error {
 		return err
 	}
 
-	if output, err := b.execCommand("git", []string{"clone", b.source}, dir); err != nil {
-		return fmt.Errorf("failed to exec 'git clone %s': %s: %v", b.source, output, err)
+	args := []string{"clone", b.source}
+
+	if b.target != "" {
```
Comment (brendandburns) [inline pkg/volume/git_repo/git_repo.go]:
```go
len(b.target) != 0 {
...
}
```

**Comment by brendandburns** [comment]
Minor comments, this LGTM.

File: pkg/api/validation/validation.go (hunk 9 above)
Comment (resouer) [inline pkg/api/validation/validation.go]: I think we allow target == "" here?

File: pkg/volume/git_repo/git_repo.go (hunk 10 above)
Comment (resouer) [inline pkg/volume/git_repo/git_repo.go]: Yes, and I 'll update the `validation.go` to keep consistency.

File: pkg/api/types.go [hunk 11]
```diff
@@ -504,7 +504,9 @@ type GitRepoVolumeSource struct {
 	// Repository URL
 	Repository string `json:"repository"`
 	// Commit hash, this is optional
-	Revision string `json:"revision"`
+	Revision string `json:"revision, omitempty"`
+	// Clone target, this is optional
+	Directory string `json:"directory, omitempty"`
```
Comment (pmorie) [inline pkg/api/types.go]: nit: no space between comma and `omitempty`

File: pkg/api/v1/types.go [hunk 12]
```diff
@@ -673,7 +673,9 @@ type GitRepoVolumeSource struct {
 	// Repository URL
 	Repository string `json:"repository"`
 	// Commit hash for the specified revision.
-	Revision string `json:"revision"`
+	Revision string `json:"revision, omitempty"`
+	// Target directory name
+	Directory string `json:"directory, omitempty"`
```
Comment (pmorie) [inline pkg/api/v1/types.go]: same nit applies here re: space

File: pkg/api/validation/validation.go (hunk 9 above)
Comment (pmorie) [inline pkg/api/validation/validation.go]: Maybe you can generalize the validations used in the downward api volume plugin here -- all the same validations should apply AFAICT

File: pkg/volume/git_repo/git_repo.go [hunk 13]
```diff
@@ -146,24 +149,39 @@ func (b *gitRepoVolumeBuilder) SetUpAt(dir string) error {
...
+		args = append(args, b.target)
+	}
+	if output, err := b.execCommand("git", args, dir); err != nil {
+		return fmt.Errorf("failed to exec 'git %s': %s: %v",
+			strings.Join(append(args[:1], args[1:]...), " "), output, err)
```
Comment (pmorie) [inline pkg/volume/git_repo/git_repo.go]: Why not just `strings.Join(args, " ")`?

File: pkg/volume/git_repo/git_repo_test.go [hunk 14]
```diff
@@ -95,13 +102,24 @@ func testSetUp(plug volume.VolumePlugin, builder volume.Builder, t *testing.T) {
...
+		repoDir = "/kubernetes"
+	}
+	expectedDirs := []string{g.GetPath(), g.GetPath() + repoDir, g.GetPath() + repoDir}
 	if len(fcmd.Dirs) != 3 || !reflect.DeepEqual(expectedDirs, fcmd.Dirs) {
 		t.Errorf("unexpected directories: %v, expected: %v", fcmd.Dirs, expectedDirs)
 	}
 }
 
 func TestPlugin(t *testing.T) {
+	testPluginWithTarget("target_dir", t)
```
Comment (pmorie) [inline pkg/volume/git_repo/git_repo_test.go]: Optional: make this a table test

If you add another case to this test, I insist :)

**Comment by pmorie** [comment]
Sorry for the late comments, but the spacing one is important to make consistent.

File: pkg/api/validation/validation.go (hunk 9 above)
Comment (resouer) [inline pkg/api/validation/validation.go]: @pmorie Really appreciate your suggestion. Can I do the clean up in another PR? just want to keep this one simple & clean.

Comment (resouer) [inline pkg/api/validation/validation.go]: Oh @pmorie  Sorry, I misunderstood your words. I though you mean to update ALL path validations in validated.go

Will be updated in this pr :)

File: pkg/volume/git_repo/git_repo_test.go (hunk 14 above)
Comment (resouer) [inline pkg/volume/git_repo/git_repo_test.go]: Not clear what's a table test ... :(

Comment (markturansky) [inline pkg/volume/git_repo/git_repo_test.go]: You have 3 tests within your single test func.

If you are testing many scenarios, a common practice is to make a map of them, like so:  https://github.com/kubernetes/kubernetes/blob/master/pkg/api/validation/validation_test.go#L252

**Comment by resouer** [comment]
@pmorie Fixed and updated, thanks!

File: pkg/api/validation/validation.go [hunk 15]
```diff
@@ -441,21 +441,25 @@ func validateHostPathVolumeSource(hostPath *api.HostPathVolumeSource) errs.Valid
...
+
+	pathErrs := validateVolumeSourcePath(gitRepo.Directory, "directory")
+	allErrs = append(allErrs, pathErrs...)
 	return allErrs
 }
 
 func validateISCSIVolumeSource(iscsi *api.ISCSIVolumeSource) errs.ValidationErrorList {
 	allErrs := errs.ValidationErrorList{}
-	if iscsi.TargetPortal == "" {
+	if len(iscsi.TargetPortal) == 0 {
```
Comment (pmorie) [inline pkg/api/validation/validation.go]: @resouer I had meant, can we please strink these changes to string tests from this PR and fix them in another PR?  It makes this PR about more than one thing and makes it tough to see the real changes to this file.

Comment (resouer) [inline pkg/api/validation/validation.go]: Yes, I'm working on this :)

File: pkg/volume/git_repo/git_repo_test.go [hunk 16]
```diff
@@ -55,106 +56,400 @@ func TestCanSupport(t *testing.T) {
...
+
+func TestPlugin(t *testing.T) {
+	gitUrl := "https://github.com/GoogleCloudPlatform/kubernetes.git"
+	revision := "2a30ce65c5ab586b98916d83385c5983edd353a1"
+	validPluginSpec := PluginSpec{
+		volPath:    "pods/poduid/volumes/kubernetes.io~git-repo/vol1",
+		pluginName: "kubernetes.io/git-repo",
+	}
+
+	scenarios := map[string]Scenario{
```
Comment (pmorie) [inline pkg/volume/git_repo/git_repo_test.go]: The style for table tests is to define scenario types inline, like so:

```go
func TestThing(t *testing.T) {
  cases := []struct{
    vol *api.Volume
    isExpectedFailure
  }

  // other stuff
}
```

In this case, go ahead and keep the `Expected` type, but make it called `expectedCommand` (more specific, and non-exported), but move the `Scenario` into an anonymous struct.  Using anonymous structs helps keep the symbol space of a package clean.

File: pkg/volume/git_repo/git_repo_test.go [hunk 17]
```diff
@@ -55,106 +56,408 @@ func TestCanSupport(t *testing.T) {
...
+
+func TestPlugin(t *testing.T) {
+	gitUrl := "https://github.com/GoogleCloudPlatform/kubernetes.git"
+	revision := "2a30ce65c5ab586b98916d83385c5983edd353a1"
+	validPluginSpec := pluginSpec{
+		volPath:    "pods/poduid/volumes/kubernetes.io~git-repo/vol1",
+		pluginName: "kubernetes.io/git-repo",
+	}
+
+	scenarios := map[string]struct {
```
Comment (pmorie) [inline pkg/volume/git_repo/git_repo_test.go]: Nit: I personally prefer to use an array here like so:

```go
cases := []struct{
  name string
  // other fields
}
```

...because:

1.  iteration order through maps is not deterministic
2.  it helps keep indentation in check

File: pkg/volume/git_repo/git_repo_test.go [hunk 18]
```diff
@@ -55,106 +56,408 @@ func TestCanSupport(t *testing.T) {
...
+			expecteds: []expectedCommand{
+				{
+					cmd: []string{"git", "clone", gitUrl},
+					dir: "",
+				},
+			},
+			pSpec:             validPluginSpec,
+			isExpectedFailure: true,
+		},
+		"invalid-plugin-name": {
```
Comment (pmorie) [inline pkg/volume/git_repo/git_repo_test.go]: invalid plugin name isn't really a scenario in scope for this test -- that's something the tests for plugin loading cover.

**Comment by pmorie** [comment]
@resouer rebase needed

**Comment by resouer** [comment]
@pmorie Done!

**Comment by pmorie** [comment]
@resouer looks like it's needed again -- ping me on slack when you push the rebase and i will review.

File: pkg/volume/git_repo/git_repo_test.go [hunk 19]
```diff
@@ -55,106 +56,356 @@ func TestCanSupport(t *testing.T) {
...
+	// Construct fake exec outputs from fcmd
+	var fackeAction []exec.FakeCommandAction
+	for i := 0; i < len(expecteds); i++ {
+		fackeAction = append(fackeAction, func(cmd string, args ...string) exec.Cmd {
+			return exec.InitFakeCmd(&fcmd, cmd, args...)
+		})
+
+	}
+	fake := exec.FakeExec{
+		CommandScript: fackeAction,
```
Comment (pmorie) [inline pkg/volume/git_repo/git_repo_test.go]: typo: fakeAction

**Comment by pmorie** [comment]
@resouer I'm sorry, but we need another rebase.  It's been a holiday week here, but I'm fully available now to get this in.  Sorry for all the rebase work, you :100: patience.

**Comment by pmorie** [comment]
@k8s-bot test this

File: pkg/api/validation/validation_test.go [hunk 20]
```diff
@@ -505,6 +506,10 @@ func TestValidateVolumes(t *testing.T) {
 	emptyMon := api.VolumeSource{RBD: &api.RBDVolumeSource{CephMonitors: []string{}, RBDImage: "bar", FSType: "ext4"}}
 	emptyImage := api.VolumeSource{RBD: &api.RBDVolumeSource{CephMonitors: []string{"foo"}, RBDImage: "", FSType: "ext4"}}
 	emptyCephFSMon := api.VolumeSource{CephFS: &api.CephFSVolumeSource{Monitors: []string{}}}
+	startsWithDots := api.VolumeSource{GitRepo: &api.GitRepoVolumeSource{Repository: "foo", Directory: "..dots/bar"}}
+	containsDots := api.VolumeSource{GitRepo: &api.GitRepoVolumeSource{Repository: "foo", Directory: "dots/../bar"}}
+	containsSpace := api.VolumeSource{GitRepo: &api.GitRepoVolumeSource{Repository: "foo", Directory: "space/  /bar"}}
```
Comment (pmorie) [inline pkg/api/validation/validation_test.go]: If you do `"space/foo  bar/zab"` in a test case here, you won't get the failure you expect.

File: api/swagger-spec/v1.json [hunk 21]
```diff
@@ -13262,6 +13261,10 @@
      "revision": {
       "type": "string",
       "description": "Commit hash for the specified revision."
+     },
+     "directory": {
+      "type": "string",
+      "description": "Target directory name"
```
Comment (pmorie) [inline api/swagger-spec/v1.json]: I think the behavior of this option should be fully spelled out here.  How about:

```
Target directory name.  Must not contain '..' or spaces.  If '.' is supplied, the volume directory will be the git repository.  Otherwise, if specified, the volume will contain the git repository in the subdirectory with the given name.
```

File: pkg/api/validation/validation.go [hunk 22]
```diff
@@ -570,23 +573,36 @@ func validateDownwardAPIVolumeSource(downwardAPIVolume *api.DownwardAPIVolumeSou
...
+	if path.IsAbs(targetPath) {
+		allErrs = append(allErrs, validation.NewForbiddenError(field, "must not be an absolute path"))
+	}
+	items := strings.Split(targetPath, string(os.PathSeparator))
+	for index, item := range items {
+		item = strings.TrimSpace(item)
+		if item == ".." {
+			allErrs = append(allErrs, validation.NewInvalidError(field, targetPath, "must not contain \"..\""))
+		}
+		if index != 0 && len(item) == 0 {
```
Comment (pmorie) [inline pkg/api/validation/validation.go]: I don't think this does what it's supposed to.  Why is index 0 special-cased here?  Also, I'm pretty sure this will let some spaces through -- for example `"foo bar"` would be allowed by this, which is contrary to the error message.  If you want to test for spaces, you have to test for the index of a space (using `[\s\xA0]`, which will also capture nonbreaking spaces) in the trimmed string.  However, I don't know that we want to disallow spaces in names.  Since this is a new condition, maybe we ought to break the question of how to handle spaces out into a different issue.

Comment (resouer) [inline pkg/api/validation/validation.go]: I think I did this in a wrong way. If a path is filled by spaces, it will be complaint as too many arguments.
So maybe I can split path by space and check length first?

File: pkg/api/validation/validation.go [hunk 23]
```diff
@@ -570,23 +573,36 @@ func validateDownwardAPIVolumeSource(downwardAPIVolume *api.DownwardAPIVolumeSou
...
+// This validate will make sure targetPath:
+// 1. is not abs path
+// 2. does not contain '..' or ' '
+// 3. does not start with '..'
+func validateVolumeSourcePath(targetPath string, field string) validation.ErrorList {
+	allErrs := validation.ErrorList{}
+	if path.IsAbs(targetPath) {
+		allErrs = append(allErrs, validation.NewForbiddenError(field, "must not be an absolute path"))
+	}
+	items := strings.Split(targetPath, string(os.PathSeparator))
```
Comment (liggitt) [inline pkg/api/validation/validation.go]: Do we want os-varient things used in API validation?

File: pkg/api/validation/validation.go [hunk 24]
```diff
@@ -570,23 +573,36 @@ func validateDownwardAPIVolumeSource(downwardAPIVolume *api.DownwardAPIVolumeSou
...
+	for index, item := range items {
+		item = strings.TrimSpace(item)
```
Comment (liggitt) [inline pkg/api/validation/validation.go]: Doesn't seem like we should be trimming spaces here… why are we ignoring data they provided?

File: pkg/api/validation/validation.go (hunk 22 above)
Comment (liggitt) [inline pkg/api/validation/validation.go]: I could see disallowing leading or trailing spaces

File: pkg/api/validation/validation.go [hunk 25]
```diff
@@ -570,23 +573,36 @@ func validateDownwardAPIVolumeSource(downwardAPIVolume *api.DownwardAPIVolumeSou
...
+		if item == ".." {
+			allErrs = append(allErrs, validation.NewInvalidError(field, targetPath, "must not contain \"..\""))
+		}
+		if index != 0 && len(item) == 0 {
+			allErrs = append(allErrs, validation.NewInvalidError(field, targetPath, "must not contain spaces"))
+		}
+	}
+	if strings.HasPrefix(items[0], "..") && len(items[0]) > 2 {
```
Comment (liggitt) [inline pkg/api/validation/validation.go]: Why no ".." prefix?

File: pkg/volume/git_repo/git_repo.go [hunk 26]
```diff
@@ -143,24 +146,39 @@ func (b *gitRepoVolumeBuilder) SetUpAt(dir string) error {
...
 	}
 
-	subdir := path.Join(dir, files[0].Name())
+	var subdir string
+
+	switch {
+	case b.target == ".":
+		// if target dir is '.', use the current dir
+		subdir = path.Join(dir)
+	case b.target != "." && len(files) == 1:
```
Comment (liggitt) [inline pkg/volume/git_repo/git_repo.go]: `b.target != "."` is already guaranteed since the first case didn't match

File: pkg/volume/git_repo/git_repo.go [hunk 27]
```diff
@@ -143,24 +146,39 @@ func (b *gitRepoVolumeBuilder) SetUpAt(dir string) error {
...
+		subdir = path.Join(dir, files[0].Name())
+	case b.target != "." && len(files) != 1:
+		return fmt.Errorf("unexpected directory contents: %v", files)
```
Comment (liggitt) [inline pkg/volume/git_repo/git_repo.go]: Should just be `default:`?

File: pkg/api/validation/validation.go (hunk 22 above)
Comment (resouer) [inline pkg/api/validation/validation.go]: `git clone https://github.com/resouer/init foo bar` is also illegal, so I suggest to trim path, then split by space and  check length, does it make sense?

File: pkg/api/validation/validation.go (hunk 25 above)
Comment (resouer) [inline pkg/api/validation/validation.go]: It's required by downward api, so I followed this 'rule', does that make sense? Or should be removed?

File: pkg/volume/git_repo/git_repo.go (hunk 27 above)
Comment (resouer) [inline pkg/volume/git_repo/git_repo.go]: Yes, I'll refactor this one

File: pkg/api/validation/validation.go (hunk 23 above)
Comment (resouer) [inline pkg/api/validation/validation.go]: But it's validating a path on the server, which is os-varient I think.

File: pkg/api/validation/validation.go (hunk 22 above)
Comment (liggitt) [inline pkg/api/validation/validation.go]: It's only illegal if we're not quoting the destination directory argument correctly, which we should absolutely be doing

File: pkg/api/validation/validation.go (hunk 23 above)
Comment (liggitt) [inline pkg/api/validation/validation.go]: The API validation shouldn't depend on which OS the API is running on, especially since the node's OS is the one that would be affected by this path

File: pkg/api/validation/validation.go (hunk 22 above)
Comment (resouer) [inline pkg/api/validation/validation.go]: I see, so just trim the path string is enough in this case?

File: pkg/api/validation/validation.go (hunk 23 above)
Comment (resouer) [inline pkg/api/validation/validation.go]: Yeah, I can see we should use the real `os.PathSeparator` here, but may I leave it to another issue? Maybe a TODO here?

File: pkg/api/validation/validation.go (hunk 22 above)
Comment (liggitt) [inline pkg/api/validation/validation.go]: Don't trim, reject. We shouldn't mutate their data to accept it, we should ensure the data can be used by the kubelet as-is
